Mathlib/CategoryTheory/Abelian/SerreClass/MorphismProperty.lean

```diff
@@ -0,0 +1,147 @@
+We show that `P.isoModSerre` is multiplicative, satisfies the two out
+of three property and is stable under retracts. (Similarly, we define
+`P.monoModSerre` and `P.epiModSerre`.)
+
+## TODO
+
+* show that a localized category with respect to `P.isoModSerre` is abelian.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+open Category Limits MorphismProperty
+
+variable {C : Type u} [Category.{v} C] [Abelian C]
+
+namespace ObjectProperty
+
+variable (P : ObjectProperty C)
+
+/-- The class of monomorphisms modulo a Serre class: given a
+Serre class `P : ObjectProperty C`, this is the class of morphisms `f`
+such that `kernel f` satisfy `P`. -/
+@[nolint unusedArguments]
+def monoModSerre [P.IsSerreClass] : MorphismProperty C :=
+  fun _ _ f ↦ P (kernel f)
+
+/-- The class of epimorphisms modulo a Serre class: given a
+Serre class `P : ObjectProperty C`, this is the class of morphisms `f`
+such that `cokernel f` satisfy `P`. -/
+@[nolint unusedArguments]
+def epiModSerre [P.IsSerreClass] : MorphismProperty C :=
+  fun _ _ f ↦ P (cokernel f)
+
+/-- The class of isomorphisms modulo a Serre class: given a
+Serre class `P : ObjectProperty C`, this is the class of morphisms `f`
+such that `kernel f` and `cokernel f` satisfy `P`. -/
+@[nolint unusedArguments]
+def isoModSerre [P.IsSerreClass] : MorphismProperty C :=
+  P.monoModSerre ⊓ P.epiModSerre
+
+variable [P.IsSerreClass]
+
+lemma monoModSerre_iff {X Y : C} (f : X ⟶ Y) :
+    P.monoModSerre f ↔ P (kernel f) := Iff.rfl
+
+lemma monomorphisms_le_monoModSerre : monomorphisms C ≤ P.monoModSerre :=
+    fun _ _ f (_ : Mono f) ↦ P.prop_of_isZero (isZero_kernel_of_mono f)
+
+lemma monoModSerre_of_mono {X Y : C} (f : X ⟶ Y) [Mono f] :
+    P.monoModSerre f :=
+  P.monomorphisms_le_monoModSerre f (monomorphisms.infer_property f)
+
+lemma epiModSerre_iff {X Y : C} (f : X ⟶ Y) :
+    P.epiModSerre f ↔ P (cokernel f) := Iff.rfl
+
+lemma epimorphisms_le_epiModSerre : epimorphisms C ≤ P.epiModSerre :=
+    fun _ _ f (_ : Epi f) ↦ P.prop_of_isZero (isZero_cokernel_of_epi f)
+
+lemma epiModSerre_of_epi {X Y : C} (f : X ⟶ Y) [Epi f] :
+    P.epiModSerre f :=
+  P.epimorphisms_le_epiModSerre f (epimorphisms.infer_property f)
+
+lemma isoModSerre_iff {X Y : C} (f : X ⟶ Y) :
+    P.isoModSerre f ↔ P.monoModSerre f ∧ P.epiModSerre f := Iff.rfl
+
+lemma isoModSerre_iff_of_mono {X Y : C} (f : X ⟶ Y) [Mono f] :
+    P.isoModSerre f ↔ P.epiModSerre f := by
+  have := P.monoModSerre_of_mono f
+  rw [isoModSerre_iff]
+  tauto
+
+lemma isoModSerre_iff_of_epi {X Y : C} (f : X ⟶ Y) [Epi f] :
+    P.isoModSerre f ↔ P.monoModSerre f := by
+  have := P.epiModSerre_of_epi f
+  rw [isoModSerre_iff]
+  tauto
+
+lemma isoModSerre_of_mono {X Y : C} (f : X ⟶ Y) [Mono f] (hf : P.epiModSerre f) :
+    P.isoModSerre f := by
+  rwa [isoModSerre_iff_of_mono]
+
+lemma isoModSerre_of_epi {X Y : C} (f : X ⟶ Y) [Epi f] (hf : P.monoModSerre f) :
+    P.isoModSerre f := by
+  rwa [isoModSerre_iff_of_epi]
```

Comment:
They’re convenient, but are they really needed? The first is `(P.isoModSerre_iff_of_mono f).mpr hf
` and the second  `(P.isoModSerre_iff_of_epi f).mpr hf`. This does not seem too painful to inline.